### PR TITLE
kernel: stack: Core thread stack usage logging

### PR DESCRIPTION
**Pull request by cvinayak**
Add stack usage logging in idle thread. Enabled by
CONFIG_INIT_STACKS and CONFIG_THREAD_MONITOR.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

### CODE REVIEW

File: kernel/idle.c [hunk 1]
```diff
@@ -151,6 +177,18 @@ void idle(void *unused1, void *unused2, void *unused3)
 	}
 #else
 	for (;;) {
+#if defined(CONFIG_INIT_STACKS)
+		if (k_uptime_get_32() - idle_ts > K_SECONDS(5)) {
+			STACK_ANALYZE("interrupt stack", _interrupt_stack);
+			STACK_ANALYZE("sys work q stack", sys_work_q_stack);
+			STACK_ANALYZE("main stack", _main_stack);
+			STACK_ANALYZE("idle stack", _idle_stack);
```
Comment (jhedberg) [inline kernel/idle.c]: Why do the above four threads get special treatment? Wont they be included in the `k_thread_foread()`?

File: kernel/idle.c [hunk 2]
```diff
@@ -151,6 +177,18 @@ void idle(void *unused1, void *unused2, void *unused3)
 	}
 #else
 	for (;;) {
+#if defined(CONFIG_INIT_STACKS)
+		if (k_uptime_get_32() - idle_ts > K_SECONDS(5)) {
```
Comment (jhedberg) [inline kernel/idle.c]: For this kind of periodic logging the period should probably be made configurable through Kconfig. E.g. if the shell is enabled one might want to disable this completely and just rely on the explicit "kernel stacks" command.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: kernel/idle.c (hunk 1 above)
Comment (cvinayak) [reply kernel/idle.c]: No, its not included... that what needs to be debugged. May the code owner for the thread monitor feature needs to intervene here.

File: kernel/idle.c (hunk 2 above)
Comment (cvinayak) [reply kernel/idle.c]: Agree. Will try to remember to fix once its figured out how to get all the threads enumerated.

File: kernel/idle.c (hunk 1 above)
Comment (carlescufi) [reply kernel/idle.c]: The above four threads might be "special" to the kernel? They are in any case essential to the RTOS so it's critical that we get their info printed

**Review by carlescufi — COMMENTED** [review]
While I like the idea, I am not sure we should pollute the idle thread with this. Maybe it would be better to provide a thread-dumping kernel call that the user can then call at will from `main()`? perhaps even a variant that never returns and uses `k_sleep()`, so that the user can write:

```
main()
{
   init_my_app()

   k_stacks();
   /* unreachable */
}
```
and then in `k_stacks()`:
```
while(1) {
  print_stack_info()
  k_sleep(K_STACKS_FREQUENCY);
}
```

**Comment by carlescufi** [comment]
Or maybe even a Kconfig option that automatically calls `k_stacks()` at the end of main:
`CONFIG_LOG_STACKS` or similar.

**Comment by nashif** [comment]
https://github.com/zephyrproject-rtos/zephyr/issues/3082 for reference

**Comment by nashif** [comment]
> While I like the idea, I am not sure we should pollute the idle thread with this.

agree.

We probably should have a stack monitoring thread that can be a bit more intelligent and send notifications only when things above a certain threshold . Does not mynewt have something like that btw?

File: kernel/idle.c [hunk 3]
```diff
@@ -120,6 +121,31 @@ void z_sys_power_save_idle_exit(s32_t ticks)
...
+extern K_THREAD_STACK_DEFINE(_main_stack, CONFIG_MAIN_STACK_SIZE);
+extern K_THREAD_STACK_DEFINE(_idle_stack, CONFIG_IDLE_STACK_SIZE);
+static u32_t idle_ts;
+
+static void stack_dump(const struct k_thread *thread, void *user_data)
+{
+	LOG_MODULE_DECLARE(kernel, CONFIG_KERNEL_LOG_LEVEL);
+
+	const char *name = k_thread_name_get((struct k_thread *)thread);
+	unsigned int size = thread->stack_info.size;
```
Comment (ceolin) [inline kernel/idle.c]: should use the same type of stack_info `u32_t`. int size can change ...

File: kernel/idle.c (hunk 2 above)
Comment (ceolin) [reply kernel/idle.c]: Yep, must be possible to configure it in build time. Make clear that this is an idle thread and there is not time guarantee.

File: kernel/idle.c (hunk 1 above)
Comment (ceolin) [reply kernel/idle.c]: it is worth to check why they are not included in `k_treadh_foreach()`

**Comment by ceolin** [comment]
> Or maybe even a Kconfig option that automatically calls `k_stacks()` at the end of main:
> `CONFIG_LOG_STACKS` or similar.

agree, both options are better than introduce a new feature under the INIT_STACKS option.

**Comment by cvinayak** [comment]
Nothing has happened since May 2019, closing this until this is picked up again for implementation.

File: kernel/idle.c (hunk 1 above)
Comment (joerchan) [reply kernel/idle.c]: @cvinayak @ceolin Not sure why they weren't included when you tried this. Could it have been related to this: https://github.com/zephyrproject-rtos/zephyr/pull/23196?

I do get them now without that change.

**Pull request closed without merging**